Socketpair fix

rennergade opened the pull request:
NaClSysSocketpair was sending a user address to RustPOSIX which is incorrect, and not using CopyOutToUser. This creates an array for the syscall and copies that out once it completes.

Please ignore that I named this branch socke**_r_**pair

rennergade committed:
socket pair ptr fix

rennergade committed:
socket pair ptr fix

kapkic reviewed (approved):
Quick fix, looks good!

youngfalcon reviewed (approved):
LGTM!

jesings reviewed (commented):
I would just like to make sure that we try to as much as possible have the failures happen before rustposix (so we don't leave orphaned file descriptors). I think you should check the address range is valid before calling NaClCopyOutToUser perhaps with a NaClUserToSysAddrRange? If not that you could do lind_close on both of the lind fds.
Otherwise this is a great and straightforward fix to a baffling problem.

rennergade committed:
socket pair ptr fix

jesings reviewed (approved):
Looks good now!

rennergade commented:
Yeah, CopyOutToUser checks address range internally, so I just added closes for the RustPOSIX fds on error. Since this is the same as how SysPipe works, I added those checks there as well.
